File: ninchatsdk/src/main/java/com/ninchat/sdk/activities/NinchatBaseActivity.java
package com.ninchat.sdk.activities;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import androidx.annotation.IdRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.annotation.StringRes;
import android.view.View;
import android.widget.TextView;

import com.ninchat.sdk.NinchatSessionManager;
import com.ninchat.sdk.R;

/**
 * Created by Jussi Pekonen (jussi.pekonen@qvik.fi) on 17/08/2018.
 */
abstract class NinchatBaseActivity extends Activity {

    protected static final int STORAGE_PERMISSION_REQUEST_CODE = "ExternalStorage".hashCode() & 0xffff;

    protected NinchatSessionManager sessionManager = NinchatSessionManager.getInstance();

    abstract protected @LayoutRes int getLayoutRes();

    protected boolean allowBackButton() {
        return false;
    }

    protected boolean hasFileAccessPermissions() {
        return checkCallingOrSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    protected void requestFileAccessPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, STORAGE_PERMISSION_REQUEST_CODE);
        }
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutRes());
    }

    protected void showError(final @IdRes int layoutId, final @StringRes int message) {
        final TextView errorMessage = findViewById(R.id.error_message);
        errorMessage.setText(message);
        findViewById(R.id.error_close).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                findViewById(layoutId).setVisibility(View.GONE);
            }
        });
        findViewById(layoutId).setVisibility(View.VISIBLE);
    }

    @Override
    public void onBackPressed() {
        if (allowBackButton()) {
            super.onBackPressed();
        }
    }
}

File: ninchatsdk/src/main/java/com/ninchat/sdk/activities/NinchatBaseActivity.kt
package com.ninchat.sdk.activities

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.ninchat.sdk.R
import com.ninchat.sdk.utils.display.getStatusBarHeight
import com.ninchat.sdk.utils.misc.Broadcast

abstract class NinchatBaseActivity : AppCompatActivity() {
    @get:LayoutRes
    protected abstract val layoutRes: Int

    protected open fun allowBackButton(): Boolean {
        return false
    }

    protected fun hasFileAccessPermissions(): Boolean {
        return checkCallingOrSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
    }

    protected fun requestFileAccessPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                STORAGE_PERMISSION_REQUEST_CODE
            )
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layoutRes)
        LocalBroadcastManager.getInstance(applicationContext).run {
            registerReceiver(closeActivityReceiver, IntentFilter(Broadcast.CLOSE_NINCHAT_ACTIVITY))
        }
    }

    override fun onDestroy() {
        LocalBroadcastManager.getInstance(applicationContext).run {
            unregisterReceiver(closeActivityReceiver)
        }
        super.onDestroy()
    }

    protected fun showError(@IdRes layoutId: Int, @StringRes message: Int) {
        findViewById<TextView>(R.id.error_message)?.run {
            setText(message)
        }
        findViewById<View>(R.id.error_close)?.run {
            setOnClickListener {
                findViewById<View>(layoutId)?.run { visibility = View.GONE }
            }
        }
        findViewById<View>(layoutId)?.run { visibility = View.VISIBLE }
    }

    override fun onBackPressed() {
        if (allowBackButton()) {
            super.onBackPressed()
        }
    }

    private val closeActivityReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent?) {
            if (intent?.action == Broadcast.CLOSE_NINCHAT_ACTIVITY) {
                setResult(RESULT_CANCELED, null)
                finish()
            }
        }
    }

    companion object {
        @JvmField
        val STORAGE_PERMISSION_REQUEST_CODE = "ExternalStorage".hashCode() and 0xffff
    }
}